rate/fetchers/nbu.go

```diff
@@ -0,0 +1,103 @@
+package fetchers
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"log/slog"
+	"net/http"
+	"slices"
+	"time"
+
+	"github.com/Hukyl/genesis-kma-school-entry/rate"
+)
+
+// NBURateFetcher is a RateFetcher implementation that fetches rates from
+// the National Bank of Ukraine
+// API docs: https://bank.gov.ua/ua/open-data/api-dev
+// NOTE: CurrencyTo can only be "UAH", as the NBU API only supports fetching rates for UAH
+//
+// Example usage:
+//
+//	fetcher := NewNBURateFetcher()
+//	rate, err := fetcher.FetchRate("USD", "UAH")
+//	if err != nil {
+//		log.Fatal(err)
+//	}
+//	fmt.Println(rate)
+type NBURateFetcher struct {
+	next RateFetcher
+}
+
+func (n *NBURateFetcher) SupportedCurrencies(_ context.Context) []string {
+	return []string{"UAH", "USD"}
+}
+
+func (n *NBURateFetcher) formatURL(cc string, date time.Time) string {
+	currentDate := fmt.Sprintf("%d%02d%02d", date.Year(), date.Month(), date.Day())
+	baseURL := "https://bank.gov.ua/NBUStatService/v1/statdirectory/" +
+		"exchange?valcode=%s&date=%s&json"
+	return fmt.Sprintf(baseURL, cc, currentDate)
+}
+
+func (n *NBURateFetcher) fetchRate(ctx context.Context, ccFrom, ccTo string) (rate.Rate, error) {
+	if ccTo != "UAH" {
+		return rate.Rate{}, fmt.Errorf("invalid currency from: %s", ccFrom)
+	}
+	result := rate.Rate{
+		CurrencyFrom: ccFrom,
+		CurrencyTo:   ccTo,
+		Time:         time.Now(),
+	}
+	if !slices.Contains(n.SupportedCurrencies(ctx), ccFrom) {
+		return result, fmt.Errorf("unsupported currency: %s", ccFrom)
+	}
+	formattedURL := n.formatURL(ccFrom, time.Now())
+	req, err := http.NewRequest(http.MethodGet, formattedURL, nil)
```

Comment:
you could add context right here, instead of doing it on line #60. Just use 
https://pkg.go.dev/net/http#NewRequestWithContext